File: 6 – Practice/leet code/009.  Palindrome Number.cpp
// https://leetcode.com/problems/palindrome-number/

class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0) {
            return false;
        }

        if (x < 10) {
            return true;
        }

        unsigned rev = 0;
        unsigned copy = x;
        
        while (copy) {
            rev = rev * 10 + copy % 10;
            copy /= 10;
        }

        return (rev == (unsigned) x);
    }
};

File: 6 – Practice/leet code/009. Palindrome Number.java
// https://leetcode.com/problems/palindrome-number/

class Solution {
    public boolean isPalindrome(int x) {
        int l = 0;
        int r = (int) (Math.log10(x) + 1);

        while (l <= r) {
            if (x < 0 || x / (int) Math.pow(10, l++) % 10 != x / (int) Math.pow(10, --r) % 10) {
                return false;
            }
        }

        return true;
    }
}
